enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,255 @@
+default [OAuthClient](https://docs.openshift.com/container-platform/4.9/rest_api/oauth_apis/oauthclient-oauth-openshift-io-v1.html)
+resources. It already provisions a client for `oc` but sets the `respondWithChallenges` to _true_ which is unsuitable
+for OAuth Grant Flow. Therefore, another client is required where the `respondWithChallenges` is set to `false` which
+should also be provisioned by the operator.
+
+### Risks and Mitigations
+
+1. There are no changes to existing behavior for users of `oc`. Previous login workflows continue working with the same
+   behavior as in previous versions.
+2. The new login workflow will need to listen on a port for the callback from the OAuth server. This port cannot be
+   fixed because the port may be in-use on the user's machine and the user may not have the permissions to resolve this.
+   Instead, the user should be able to specify an alternate port which is free on the system. In the new client we will
+   allow redirects on the following URLs:
+   * `http://localhost:8280/callback`
+   * `http://localhost:8281/callback`
+   * `http://localhost:8282/callback`
+   * `http://localhost:8283/callback`
+   * `http://localhost:8284/callback`
+3. If the user wants to customize the callback URL further they should create another `OAuthClient` with a different
+   client ID and redirect URL.
+4. Some OSes under certain configurations might prevent the callback server from listening on the callback port. In this
+   case the server would not start and `oc` would print a message that it cannot listen on that port.
+5. If `localhost` on the user's machine is not the machine itself but has been compromised and points to another
+   location then the attacker can obtain the security code. But this code cannot be exchanged for a token because the
+   attacker does not have access to the challenge key which is only in memory on the users machine.
+
+## Design Details
+
+### Open Questions
+
+1. Should we modify the OAuth server library that it allows any port for the redirect URL when the url is localhost?
+2. Does receiving a callback on `localhost` pose a security risk? Or in other words what happens when `localhost` on the
+   users machines actually points to a location other than the machine itself.
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+- What additional testing is necessary to support managed OpenShift service-based offerings?
+
+No need to outline all of the test cases, just the general strategy. Anything that would count as tricky in the
+implementation and anything particularly challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage expectations).
```

Comment:
I expect this section to be filled.

Reply:
I've added some details about the tests.